src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+{
+	unsigned int     mode     = 0664;
+	int              two_args = 1, rc, is_target_path, idx_fd, idx_dirfd;
+	char             item_name[DFS_MAX_NAME];
+	char             parent_dir[DFS_MAX_PATH];
+	char             full_path[DFS_MAX_PATH];
+	dfs_obj_t       *dfs_obj;
+	dfs_obj_t       *parent;
+	mode_t           mode_query = 0, mode_parent = 0;
+	struct dfs_mt   *dfs_mt;
+
+	if (pathname == NULL) {
+		errno = EFAULT;
+		return (-1);
+	}
+	if (oflags & O_CREAT) {
+		va_list arg;
+
+		va_start(arg, oflags);
+		mode = va_arg(arg, unsigned int);
+		mode = mode & mode_not_umask;
+		va_end(arg);
+		two_args = 0;
+	}
+
+	if (!hook_enabled)
+		goto org_func;
+
+	rc = parse_path(pathname, &is_target_path, &parent, item_name, parent_dir,
+			full_path, &dfs_mt);
+	if (rc != 0 && errno == ENOENT)
+		D_GOTO(out_error, rc = ENOENT);
+
+	if (!is_target_path)
+		goto org_func;
+
+	if (oflags & __O_TMPFILE) {
+		if (!parent && (strncmp(item_name, "/", 2) == 0))
+			rc = dfs_access(dfs_mt->dfs, NULL, NULL, X_OK | W_OK);
+		else
+			rc = dfs_access(dfs_mt->dfs, parent, item_name, X_OK | W_OK);
+		if (rc) {
+			if (rc == 1)
+				rc = 13;
+			D_GOTO(out_error, rc);
+		}
+	}
+
+	if ((oflags & O_RDWR) && (oflags & O_CREAT)) {
+		mode_parent = dfs_obj_mode(dfs_mt->dfs, parent);
+		if ((S_IXUSR & mode_parent) == 0 || (S_IWUSR & mode_parent) == 0)
+			D_GOTO(out_error, rc = EACCES);
+	}
+	/* file/dir should be handled by DFS */
+	if (oflags & O_CREAT) {
+		rc = dfs_open(dfs_mt->dfs, parent, item_name, mode | S_IFREG, oflags, 0, 0, NULL,
+			      &dfs_obj);
+		mode_query = S_IFREG;
+	} else if (!parent && (strncmp(item_name, "/", 2) == 0)) {
+		rc = dfs_lookup(dfs_mt->dfs, "/", oflags, &dfs_obj, &mode_query, NULL);
+	} else {
+		rc = dfs_lookup_rel(dfs_mt->dfs, parent, item_name, oflags, &dfs_obj, &mode_query,
+				    NULL);
+	}
+
+	if (rc)
+		D_GOTO(out_error, rc);
+
+	if (S_ISDIR(mode_query)) {
+		idx_dirfd = find_next_available_dirfd(NULL);
+		assert(idx_dirfd >= 0);
+
+		dir_list[idx_dirfd]->dfs_mt      = dfs_mt;
+		dir_list[idx_dirfd]->fd          = idx_dirfd + FD_DIR_BASE;
+		dir_list[idx_dirfd]->offset      = 0;
+		dir_list[idx_dirfd]->dir         = dfs_obj;
+		dir_list[idx_dirfd]->num_ents    = 0;
+		dir_list[idx_dirfd]->st_ino      = FAKE_ST_INO(full_path);
+		memset(&dir_list[idx_dirfd]->anchor, 0, sizeof(daos_anchor_t));
+		if (strncmp(full_path, "/", 2) == 0)
+			full_path[0] = 0;
+
+		/* allocate memory for path and ents. */
+		D_ALLOC(dir_list[idx_dirfd]->path, (sizeof(struct dirent)*READ_DIR_BATCH_SIZE +
+			DFS_MAX_PATH));
+		if (dir_list[idx_dirfd]->path == NULL) {
+			free_dirfd(idx_dirfd);
+			D_GOTO(out_error, rc = ENOMEM);
+		}
+		dir_list[idx_dirfd]->ents = (struct dirent *)(dir_list[idx_dirfd]->path +
+					   DFS_MAX_PATH);
+		rc = snprintf(dir_list[idx_dirfd]->path, DFS_MAX_PATH, "%s%s", dfs_mt->fs_root,
+			      full_path);
+		if (rc >= DFS_MAX_PATH) {
+			printf("Path is longer than DFS_MAX_PATH\nQuit.\n!");
+			exit(1);
+		}
```

Comment:
This is odd but I think I've seen what you've done.  It would be much safer to use D_ALLOC_ARRAY and D_ASPRINTF() instead and it would use less memory as well.

Reply:
I was trying to minimizing the number of calls of malloc(). ok. I just updated as you suggested. Thank you!